### DISCUSSION TITLE
Riverpod in a custom package/library

### DISCUSSION THREAD

**Discussion by bytcra** [rrousselGit/riverpod#2569] [Q&A]
I m creating a package/library and want to implement riverpod as state management. Today I get to see the real shortcoming of Google of not understanding the context. When I try to search "Using riverpod for custom package or library" it gives me results of implementing riverpod the general way. 

Anyway, I tried to find the same on the discussion thread and I might have missed out or something but I did not find the answer.

My question:
How to implement riverpod 2.0 the right way in a package where there is no main() to start with?

**Comment by rrousselGit** [comment]
It widely depends on the package you want to make. Hard to say without giving your constraints.

For example, you could make a package exposing a bunch of providers. Or you could expose a bunch of widgets internally using providers, and ask for folks to wrap their app in ProviderScope (or a custom-made widget that internally uses ProviderScope)

**Reply by bytcra** [reply]
Thank you for the reply.

This will mandate the user to install riverpod too, right?  I have direct screens made of stateful widgets exposed.. what would be the best way to setup a provider scope without user needing to install riverpod in their app.

**Reply by rrousselGit** [reply]
You'll need a custom wrapper widget placed above the entire app, which internally uses ProviderScope

**Reply by bytcra** [reply]
Something like this?

```dart
class MyWidgetWrapper extends StatelessWidget {
  const MyWidgetWrapper({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return ProviderScope(child: MyWidget());
  }
}

``` 
or

```dart
class MyWidgetWrapper<T> extends StatelessWidget {
  const MyWidgetWrapper({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return ProviderScope(child: T as ConsumerWidget);
  }
}
``` 

though the second one gives me:
type '_Type' is not a subtype of type 'ConsumerWidget' in type cast

**Reply by akshaydoshi2** [reply]
@bytcra did you find a solution?
I am trying to implement Riverpod in a custom package but I am stuck at the same point.

**Comment by jordanb** [comment]
Maybe this can help => https://github.com/rrousselGit/riverpod/discussions/1613#discussioncomment-10603619 ?

**Comment by nicolasbraun** [comment]
Hello, just checkin in one, I have the same requirements and could not find anywhere a proper response.

 I'm building a chat lib whose internal components (bubble etc ..) use riverpod to access provider created at the root level (for example the `ChatTheme` 

- riverpod needs a ProviderScope to work
- I am not sure I can use a nested one. How would it behave if I add a `ProviderScope` to an app which already has one? My users can overide part of the widgets in my lib-tree so they must be able to user `Consumer`s if they use Riverpod as state management.

These lines in the code  make me think if I do not override anything it should work but I'm not sure the provider defined in my app will be available in my  sub-`ProviderScope` ?
https://github.com/rrousselGit/riverpod/blob/5a513c5aeee5a91d7cf5aa9a6e6a414b822bdee9/packages/flutter_riverpod/lib/src/framework.dart#L47


If not  those options but would like you 2 cents on it @rrousselGit 

• Ask user to wrap their app in a custom widget, something like
```dart
class ChatProvider extends StatelessWidget {
  final Widget child;

  const ChatProvider({required this.child, super.key});

  @override
  Widget build(BuildContext context) {
    returnProviderScope(child: child); 
  }
}
```

• Wrap my main lib component in some kind of ProviderScope checker
```dart
class SafeChatWidget extends StatelessWidget {
  final Widget child;

  const SafeChatWidget({required this.child, super.key});

  @override
  Widget build(BuildContext context) {
    return ProviderScope.containerOf(context, listen: false, nullOk: true) != null
        ? child
        : ProviderScope(child: child);
}
````

Thanks
